src/OrchardCore.Modules/OrchardCore.Secrets/Services/SecretHybridDecryptor.cs

```diff
@@ -0,0 +1,48 @@
+using System;
+using System.IO;
+using System.Security.Cryptography;
+using OrchardCore.Secrets.Models;
+
+namespace OrchardCore.Secrets.Services;
+
+public class SecretHybridDecryptor : ISecretDecryptor
+{
+    private readonly SecretHybridEnvelope _envelope;
+    private readonly RSASecret _encryptionSecret;
+    private readonly RSASecret _signingSecret;
+
+    public SecretHybridDecryptor(SecretHybridEnvelope envelope, RSASecret encryptionSecret, RSASecret signingSecret)
+    {
+        _envelope = envelope;
+        _encryptionSecret = encryptionSecret;
+        _signingSecret = signingSecret;
+    }
+
+    public string Decrypt(string protectedData)
+    {
+        var protectedBytes = Convert.FromBase64String(_envelope.ProtectedData);
+        var signatureBytes = Convert.FromBase64String(_envelope.Signature);
+
+        // The private key has been used for signing, the matching public key should be used for verification.
+        using var rsaSigner = RSAGenerator.GenerateRSASecurityKey(2048);
+        rsaSigner.ImportRSAPublicKey(_signingSecret.PublicKeyAsBytes(), out _);
+        if (!rsaSigner.VerifyData(protectedBytes, signatureBytes, HashAlgorithmName.SHA256, RSASignaturePadding.Pkcs1))
+        {
+            throw new CryptographicException("Could not verify signature.");
+        }
+
+        // The public key has been used for encryption, the matching private key should be used for decryption.
+        using var rsaDecryptor = RSAGenerator.GenerateRSASecurityKey(2048);
+        rsaDecryptor.ImportRSAPrivateKey(_encryptionSecret.PrivateKeyAsBytes(), out _);
+        var aesKey = rsaDecryptor.Decrypt(Convert.FromBase64String(_envelope.Key), RSAEncryptionPadding.Pkcs1);
+
+        using var aes = Aes.Create();
```

Comment:
It's probably not a deal breaker here (as chances of seeing encrypted secrets in the DB being altered are probably low), but it's worth noting this encryption/decryption routine is prone to chosen-ciphertext attacks: one could theoretically manipulate the ciphertext in a way that the decrypted value wouldn't match the expected value and be tampered with.

To fix that, moving to an authenticated encryption algorithm - like AES-GCM or AES-CBC with some additional HMAC protection (e.g encrypt-then-MAC: https://en.wikipedia.org/wiki/Authenticated_encryption#Encrypt-then-MAC_(EtM)) - will be required.

I'm sure I mentioned it earlier in this PR, but it's worth exploring built-in options first to see if we can't leverage existing stuff instead of rolling our own crypto. Good candidates include:
  - XML Encryption/Signing (supported by the BCL via `EncryptedXml` and `SignedXml`).
  - JSON Web Encryption (JWE)/JSON Web Signing (JWS): it isn't supported by the BCL - IdentityModel supports these standards, but is designed to produce JSON Web Tokens and isn't really suitable for signing or encrypting arbitrary data - but there are good third-party options like JOSE-JWT that are more flexible for encrypting/signing arbitrary data: https://github.com/dvsekhvalnov/jose-jwt

Reply:
Scratch the first part of my message... I missed there was a RSA signature applied to the ciphertext 🤣

Reply:
Okay but I saw below that you still recommend to use a JSON Web Encryption/Signing.